### ISSUE TITLE
[A11Y] Add the HTML details element to core form API

### ISSUE THREAD

**Issue by jenlampton** [backdrop/backdrop-issues#324]
Drupal 8 decided to replace fieldsets with details, see [Freedom For Fieldsets! Long Live The DETAILS.](https://www.drupal.org/node/1168246)

Here's the change record for the Drupal issue: https://www.drupal.org/node/1852020

Switching all Fieldsets has caused some significant issues for Drupal, so instead of replacing existing fieldsets, we probably should provide `details` as _additional_ Form API element.

Here's an example for that element on [developer.mozilla.org](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details)

This issue is a blocker for 
* https://github.com/backdrop/backdrop-issues/issues/5090

---

PR: https://github.com/backdrop/backdrop/pull/3408

**Comment by stpaultim** [comment]
@jenlampton - Its six years later and this doesn't seem to be an issue. Do we need to have this ticket open to discuss the reasons why we should maybe not do something that we appear pretty comfortable not doing. :-)

I'm going out on a limb and closing it. Please reopen if that was a mistake.

**Comment by klonos** [comment]
Well, fieldsets are a Drupalism; whereas details elements are  web standard, which gets usability improvements etc., so there is merit in the change. But since the change is likely to break existing themes, if we decide to reopen this, then it should be slated for 2.x.

**Comment by klonos** [comment]
...here's another related documentation issue in d.org: https://www.drupal.org/project/documentation/issues/1892968

**Comment by indigoxela** [comment]
While replacing all fieldsets at once seems impossible, it could be still feasible to *add* `details/summary` elements to our form API - just an additional option.

Modules and core can then replace fieldsets step by step wherever appropriate.
(I don't think that fieldsets are a Drupalism, btw, but `details` seem more lightweight (no js) and the browser support looks good now.)

Or could this make Backdrop look weird?

**Comment by klonos** [comment]
> Modules and core can then replace fieldsets step by step wherever appropriate.

👍 

> I don't think that fieldsets are a Drupalism

Right, the collapse/expand functionality is though. Whereas the details element has in-built support for that.

**Comment by indigoxela** [comment]
@stpaultim it seems that trying to close that issue actually revived it. :smile: 

I see some interest in having `'#type' => 'details'` in core form API. But we don't want to rip fieldsets out, we just want to add `details` as available form items wrapper.

The issue title needs some update and so does the description.

**Comment by laryn** [comment]
I support adding `details` as a new form API option.

Here's a demo for those unfamiliar: https://codepen.io/denic/pen/PozobRO

**Comment by jenlampton** [comment]
Sounds like that's consensus :)

**Comment by stpaultim** [comment]
> @stpaultim it seems that trying to close that issue actually revived it.

I love it!

**Comment by indigoxela** [comment]
Here's a [PR](https://github.com/backdrop/backdrop/pull/3408) as a starting point. It still needs work, it currently provides the *minimum* code needed to use the `details` element on the status page (admin/reports/dblog).

Feedback is welcome.

**Comment by indigoxela** [comment]
Before I continue working on the PR, I'd like to see a decision **where** in core fieldsets should get replaced by details.

If I search for `'#collapsible' => TRUE` in core I get more than 90 results. If we replaced all of those at once, the PR would become *huge* (and very hard to review). (And very likely loads of tests to fix...)

Any ideas for a possible subset for the initial PR to prevent that?

**Comment by klonos** [comment]
> Any ideas for a possible subset for the initial PR to prevent that?

`admin/reports/dblog` is a good start for initial testing, but in the PR sandbox both fieldsets are not rendered at all 🤷 . Another place to use for testing could be the global field settings (`admin/structure/types/manage/%/fields/%`).

Another idea (although not FAPI-related) would be to replace the more/less links in the module listing page (module tags and dependencies) and the status report (for errors/warnings). Separate issue for this?

One thing to note is that D8 implementation seems to have used `#title` (see https://www.drupal.org/node/1852020), whereas your PR @indigoxela uses a new `#summary` property. Should we also stick with `#title` for consistency? ...or support both of them interchangeably (transparently, for better DX when porting modules or reusing Drupal code snippets)?

**Comment by indigoxela** [comment]
> in the PR sandbox both fieldsets are not rendered at all

WTF? They were there two days ago. Now I get "Theme hook "details" not found." Oops, after flushing caches it's also gone on my local dev.

I suspect, my "minimal code" is yet too minimal. I'll fix that as soon as I find the time.

> ...Another idea...

@klonos please do not expand this issue even more. :wink: It will already be a big one even without additional requests.

> One thing to note is that D8 implementation seems to have used #title 

I noticed that, but suppose it's still #title in D8, because they simply overwrote the original fieldset code (I checked some of their related commits). We add something new. IMO it makes sense to use the actual name of the html tag wrapping the values. Correct me if I'm wrong, though. I have no problem with using `#title` and `#description` instead of `#summary` and `#details`.

**Comment by indigoxela** [comment]
> in the PR sandbox both fieldsets are not rendered at all

Hehe, so silly, :facepalm:  found and fixed it. The sandbox renders details now on dblog page.

**Comment by olafgrabienski** [comment]
I've had a look at the sandbox website (admin/reports/dblog). The details element works fine. It's also a visual improvement as -- in contrast to the current fieldsets on that page -- it's not flickering.

Idea for a follow-up: improve the look and feel further, it feels still not ideal, e.g. cursor, outline (for reference: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/details#Customizing_the_appearance)

**Comment by docwilmot** [comment]
> Before I continue working on the PR, I'd like to see a decision where in core fieldsets should get replaced by details.

Just to be clear, I thought that we were simply adding this as a Form API *option* rather than actually implementing in core? 

The browser support seems solid, so I personally feel we should use it. But if we implement it anywhere, we should implement it everywhere. Its either fit to replace fieldsets or it isnt.

**Comment by indigoxela** [comment]
> Idea for a follow-up: improve the look and feel further, it feels still not ideal

The styles are currently minimal - just to make the details element look alike the fieldsets (no perfect fit, I admin). Suggestions are welcome.

> I thought that we were simply adding this as a Form API option rather than actually implementing in core?

We will have to implement it *somewhere* to add functional tests. And just adding without using it seems a bit weird.
_Collapsible_ fieldsets are used all over the place - by almost every core module. If we replace it everywhere, the change is huge and the PR will be hard to review - that's my main concern. I'm leaning towards baby-steps.

To be clear: I wouldn't recommend to replace fieldsets that are simple wrappers (not collapsible, just normal fieldsets), because the details element wouldn't make much sense there (D8 did, but it feels odd, details are always collapsible).

**Comment by klonos** [comment]
> Its either fit to replace fieldsets or it isnt.

It can only replace collapsible (`'#collapsible' => TRUE`) fieldsets - it is NOT fit for always-expanded ones, as there is no way to "lock" the HTML `<details>` element to always be expanded (at least I could not find a way to do it).

**Comment by klonos** [comment]
> Hehe, so silly, 🤦 found and fixed it. The sandbox renders details now on dblog page.

Yup, things work as expected now 👍 ...the only thing that I find annoying is the focus ring when clicking the legend.

> @klonos please do not expand this issue even more. ...

Yeah, as I said, that can be a follow-up.

**Comment by klonos** [comment]
> It's also a visual improvement as -- in contrast to the current fieldsets on that page -- it's not flickering.

I almost forgot what has triggered this issue to revive: this "flickering" when expanding/collapsing the "Format options" fieldset below each CKEditor instance. @indigoxela I would like that to be fixed as part of this issue here if possible please.

**Comment by indigoxela** [comment]
> ...the only thing that I find annoying is the focus ring when clicking the legend.

Yikes! Webkit... fixed by latest commit.

**Comment by klonos** [comment]
> Yikes! Webkit... fixed by latest commit.

Fixed indeed, but you have unconditionally removed the ring in ALL cases. we need to keep the focus when using keyboard though, for accessibility.

**Comment by indigoxela** [comment]
> ... this "flickering" when expanding/collapsing the "Formatting options" fieldset below each CKEditor instance. I would like that to be fixed as part of this issue here if possible please.

Hey, that could be our starting point - and it would be minimal. I agree, that's a really annoying thing.

> ...we need to keep the focus when using keyboard though, for accessibility.

Is that possible? How?

**Comment by indigoxela** [comment]
I've updated the PR:

- Keep focus highlighting for keyboards
- Add general details styles to Basis theme
- Use the details element for the filter wrapper (node form)
- Use the details element for the expanded link options (CKEditor link dialog, another jumpy thing)

That means that the filter module doesn't use collapsible fieldsets anymore.

_Note:_ I still kept the "example" usage on dblog page (admin/reports/dblog) for now, so you can compare the old and new appearance.
We'll possibly revert that before merging, so the only usage of the new element is in the Filter module.

Only ten files changed - that's still manageable. :wink:

**Comment by docwilmot** [comment]
If its too much work, I understand. But really if we just swapped **all** the collapsible fieldsets it would still be smaller than a whole lot of previous PRs. I don't think we should limit the PR based on its size.

Also, how do we manage `'#collapsible' => TRUE, '#collapsed' => FALSE`?

**Comment by klonos** [comment]
> I've updated the PR:

Thanks @indigoxela 🙏 ...things are starting to look/work really nice 👍 

> Is that possible? How?
> 
> Update: never mind, found it.

I always had faith in you 😉 ...this works as expected with keyboard and mouse now, and the ring is nicely sized, so that it's not "tightly wrapped" around the legend text as it was previously. Good job there 👍 

Do you think we could also include the fieldsets in `admin/appearance/settings/basis` in this PR? (they also suffer from the "jumpy" animation).

>I don't think we should limit the PR based on its size.

Not sure how many changes we'd need to make in core 🤷  ...if it's a dozen or two, then I agree; if more than that, then perhaps we should be doing follow-ups once this is in.

**Comment by indigoxela** [comment]
> If its too much work, I understand. But really if we just swapped all the collapsible fieldsets ... I don't think we should limit the PR based on its size.

The problem is that there are 45 files in core that would need an update. That's hard for the testers (too many places to check) and even harder for reviewers.

> how do we manage '#collapsible' => TRUE, '#collapsed' => FALSE?

Easy: `'#open' => TRUE`. See it in action in the sandbox on the dblog page when you apply filters. :wink: 

> Do you think we could also include the fieldsets in admin/appearance/settings/basis in this PR?

Seems feasible, if people agree. Requires changes in themes/basis/theme-settings.php.

**Comment by findlabnet** [comment]
When the CKEditor link dialog is expanded, the expanded form is not fully available in the viewport (at least in Firefox).

![details](https://user-images.githubusercontent.com/11479969/100197093-81db3900-2f02-11eb-8df0-a2227ef09af2.png)

**Comment by indigoxela** [comment]
> When the CKEditor link dialog is expanded, the expanded form is not fully available in the viewport (at least in Firefox).

Confirmed. But only in Seven. With animated fieldset the dialog "jumps", but resizes to fit the viewport. With details (which expand faster) it does not jump, but the submit button stays outside the viewport.

**Comment by docwilmot** [comment]
> The problem is that there are 45 files in core that would need an update. That's hard for the testers (too many places to check) and even harder for reviewers.

We've done much bigger in past.

But question is do these files need updating or not? At some point we're going to have to dive in and do this. Will we do 5 PRs in batches of 10 files each? Whats the plan?

**Comment by klonos** [comment]
> Will we do 5 PRs in batches of 10 files each? Whats the plan?

We can do either one module each time, or admin "sections" per time I guess.

**Comment by indigoxela** [comment]
> But question is do these files need updating or not?

45 files in core use `'#collapsible' => TRUE`, many of them several times. So, yes, all of these will need updating.

My suggestion still is: do it in steps to prevent possible regressions. If we really change each and every usage of collapsible fieldsets at once, my concern is that something might slip through.

> We can do either one module each time, or admin "sections" per time I guess.

That seems like a good approach - both suggestions.

**Comment by indigoxela** [comment]
I had to revert the usage of details element inside dialog. That's a job for someone with better Javascript or jQuery UI skills.

But I changed fieldsets to details on admin/appearance/settings/basis (the Basis theme settings).

Here's an example for both, fieldset and details side by side (only to demonstrate the similar look, didn't keep it in my PR):

![details-fieldset](https://user-images.githubusercontent.com/842066/100249649-1498c980-2f3d-11eb-8692-4db24a7103cc.png)

Of course, when expanding/collapsing the difference is obvious (flickering animation vs. no animation).

**Comment by klonos** [comment]
Thanks @indigoxela ...looks good in `admin/appearance/settings/basis` 👍

**Comment by indigoxela** [comment]
Automated tests have been added. Please test and/or review and/or provide feedback.

**Comment by jenlampton** [comment]
I'd like to reiterate that we cannot change HTML (or CSS) in 1.x because it will break all sites that have styled the previous markup, or modified the previous CSS.

> Just to be clear, I thought that we were simply adding this as a Form API option rather than actually implementing in core?

This is the only safe way to approach this problem in 1.x.  If any of you have add-ons and would like to replace fieldsets with the details element, that's the place to do it until 2.x comes out. 

> We will have to implement it somewhere to add functional tests. 

The way we do this in core is by adding an additional module that will only be used for tests. In the test we can enable the module, visit the form it provides with the element, and test there.

**Comment by klonos** [comment]
> I'd like to reiterate that we cannot change HTML (or CSS) in 1.x because it will break all sites that have styled the previous markup, or modified the previous CSS.

I hate to admit it (because I would love this change to get in sooner), but you are absolutely right @jenlampton ...switching `<fieldset>`s to `<details>` does alter the rendered HTML, and this is a recipe for breaking contrib 😓 ...I have some ideas about implementing this is a backwards-compatible way, but will need to think it through a bit more.

**Comment by klonos** [comment]
...I was thinking that perhaps we could:

- add a "Use `<details>` instead of `<fieldset>`" checkbox, either globally at `admin/config/administration`, or under `admin/appearance/settings/seven` (introduce a "Settings" form for the Seven theme specifically)
- this setting would be on OOTB for new sites, but off for existing sites
- check for this setting in `theme_fieldset()`, and conditionally render the fieldset as `<details>` if enabled.

Thoughts?

**Comment by indigoxela** [comment]
> I'd like to reiterate that we cannot change HTML (or CSS) in 1.x because it will break all sites that have styled the previous markup, or modified the previous CSS.

OK, then let's close this issue as wontfix. @jenlampton you opened it, feel free to close it. And please don't waste more of my time with issues you will block anyway.

**Comment by olafgrabienski** [comment]
> OK, then let's close this issue as wontfix.

Why? The starting point of the issue was to discuss if we want to change elements from *fieldsets* to *details*, then we reached consensus to only add *details* to the form API without replacing *fieldsets*. At a certain point we also talked about replacing some fieldsets, mainly for testing purposes, and now we're here again: Add *details* to form API. Plus a suggestion how to test without implementing the change in core. That doesn't sound so bad, im my opinion.

**Comment by indigoxela** [comment]
>> OK, then let's close this issue as wontfix.
> Why?

The problem I see with code that's there, but may not be used before 2022 (or 2015) is, that it will be forgotten about. Sort of "dead code".

Personally, I see no chance to wrap it all to get a user interface to choose between fieldsets and details. At least, not without horribly bloating core code. Maybe someone else wants to take over.

**Comment by laryn** [comment]
This is getting a little far afield, but what about a general "experimental mode" button that could be set once to allow changes like this before 2025? Rather than separate individual checkboxes targeted to specific code-level changes that would be cryptic to most users.

**Comment by docwilmot** [comment]
Lets just add this as a Form API option, as per the original plan. Contrib can use it, experimentally or otherwise. Site builders can use it in entity forms. It wouldnt be forgotten.

**Comment by olafgrabienski** [comment]
What about new core elements: Could they also use it?

**Comment by jlfranklin** [comment]
Why does this need to go into core at all?  Build it as a contrib module, include a `hook_form_alter()` to update existing forms to use `details` (with an allow/deny list to include/exclude certain forms), and merge that in 2.x core when the time comes. Sites that want the `details` element can include this module and deal with any breakage in their own themes.

Contrib is the best place for "experimental mode" stuff.

**Comment by docwilmot** [comment]
> Why does this need to go into core at all? Build it as a contrib module, include a hook_form_alter() to update existing forms to use details (with an allow/deny list to include/exclude certain forms), and merge that in 2.x core when the time comes. Sites that want the details element can include this module and deal with any breakage in their own themes.

Makes sense

**Comment by jlfranklin** [comment]
Done.  [Details HTML5](https://github.com/backdrop-contrib/details_html5) does little more than repackage the PR above.

**Comment by indigoxela** [comment]
I'll try to get this issue back on track - as I definitely helped a lot to derail it. :smirk: 

We can **not** replace any of the fieldsets in core with details before 2025, but we *could* use it for new form items - if any.
And contrib modules could start to use details right away.

@jlfranklin many thanks for taking the initiative to move it all to contrib. But I think it might be better if core provides the element type. A hook_form_alter() to replace core fieldsets should, of course, be something in contrib.

Proposal for next steps:

1. People test and review the PR as it is now
2. Then I remove all usages of details in core (revert dblog.admin.inc, filter.module, theme-settings.php)
3. Another code review
4. Merge (1.18)
5. Add to our [FAPI documentation](https://api.backdropcms.org/form_api)

@jenlampton I understand that it's not possible to *change* CSS, but it is possible to *add* some that hasn't been there before and that doesn't affect existing markup - is that correct?

**Comment by klonos** [comment]
**I think that the FAPI element should be provided by core**. Contrib can override fieldsets with details and provide an option for that.

**Comment by jlfranklin** [comment]
While I can't find a citation for it, I thought it was general Drupal policy to not include code in core unless core uses it, or there is some strong technical reason why it can't be put in contrib.  It was a sensible policy that kept code that could not be exercised by a stock install from getting into core.  As a side-effect, I think it helped create the thriving contrib community that makes Drupal such a strong platform.

> I think that the FAPI element should be provided by core.

I see this sentiment often in the issue queues, and I think it explains why Backdrop has only 729 projects (526 with a release), when [Drupal has over 20,000](https://www.drupal.org/project/usage?page=212), meanwhile the lines-of-code in Backdrop has grown by a third since 1.1.0.  Our first thought is "add it to core," when it should be "can we write a contrib module for this?"

**Comment by klonos** [comment]
Chicken-egg situation I believe @jlfranklin ...my understanding is that the intention is to add this in core (eventually), but because the changes are breaking backwards compatibility, we need to hold off till v2.0. When we start working on 2.x, we'll need to add it to core, and remove/deprecate the contrib project.

> Our first thought is "add it to core," when it should be "can we write a contrib module for this?"

If that were the case, then things like CCK and Views would still be in contrib (and other HTML5 FAPI elements for that matter, like Date). Besides, this is not some random/arbitrary feature that is sporadically being used in a single place in the UI; it'll be affecting all fieldsets and their style being used everywhere in core and contrib (for the better IMO).

In Drupal, people were (and some still are) advocating for "mini core", but I thought that Backdrop is meant to make things easier, and the OOTB experience better/faster for everyone, so that they don't have to go hunting and installing contrib unless they need something very specific (the "~20% use case"). I personally don't see replacing the Drupalism that fieldsets are as a thing that will only benefit 20% or less, but that may just be me 🤷

**Comment by jenlampton** [comment]
> This is getting a little far afield, but what about a general "experimental mode" button that could be set once to allow changes like this before 2025?

This is what contrib is for :) Thank you @jlfranklin for creating the contrib module.

> I understand that it's not possible to change CSS, but it is possible to add some that hasn't been there before and that doesn't affect existing markup - is that correct?

I think so. Since there was no styling for details in core before, it seems acceptable to add some (but we should confirm with core committers). 

The only way adding new CSS could be a problem is if a site had used the contrib details_html5 module (or similar) and styled the `details` element. If that were the case, and if core provided new styles that were not completely overwritten by those previously in place, there might be some breakage.

> While I can't find a citation for it, I thought it was general Drupal policy to not include code in core unless core uses it

This is the general rule-of-thumb for Backdrop too, but in this case we have a good reason not to use it in core (backwards compatibility) and we can solve the issue of test coverage by using it in a testing module. 

> What about new core elements: Could they also use it?

Yep! That's the goal. Adding new things could be done the new way, then when 2.x rolls around we would only need to replace the old things.

> Our first thought is "add it to core," when it should be "can we write a contrib module for this?"

I agree that we do a lot to Backdrop core, and maybe it is more than we should be doing. But I also agree that we have made a substantial improvement in the OOB experience from Drupal 7. 

If we decide not to add the details element in 1.x I would like to leave the issue open - but tag it with a `2.x` milestone so that we can make the change then.

**Comment by indigoxela** [comment]
> While I can't find a citation for it, I thought it was general Drupal policy to not include code in core unless core uses it, or there is some strong technical reason why it can't be put in contrib.

Sure, we try to only add stuff to core we really need. But IMO that's the case with details elements. We want it in core, but unfortunately can't replace the collapsible fieldsets yet in 1.x - for compatibility reasons.
We will replace them in 2.x. If the form elements are already in core then, it'll be easier - less juggling with deprecation and with possible conflicts with contrib projects.

> I think it explains why Backdrop has only 729 projects (526 with a release), when Drupal has over 20,000

To be honest, I'm not convinced, that's the actual reason. And I'd like to point out that these 20k modules are for at least 3 different Core versions, aggregated over decades, many of them abandoned / unmaintained.

> In Drupal, people were (and some still are) advocating for "mini core", but I thought that Backdrop is meant to make things easier, and the OOTB experience better/faster for everyone

It can always be argued, if this or that feature should go to core or to contrib. Opinions may vary, but most of the time we're able to find consensus (I admit, sometimes we aren't). Sure, sometimes it's easier and faster to make progress in contrib, but sometimes it's just an unnecessary detour.

Why I (sort of) advocate for details in core?

I think they do a much better job than collapsible fieldsets. They're cleaner and more lightweight and don't require Javascript.

Another thought: If we're able to find consensus within the next few weeks, details could be added in the next minor release. Feature freeze for 1.18 is in a few weeks, scheduled release on January 15th.

...but I don't want to rush you (or do I? :grin:).

**Comment by klonos** [comment]
I agree with everything @indigoxela said ^^ 👍

**Comment by docwilmot** [comment]
Just to support but qualify @jlfranklin's comment on core vs contrib, we really have been very agreeable to core additions over the years (perhaps to much so). I think though most of this has been under the intention towards D8 or WP parity the vast majority of the times. We have also tried to mention, in most situations, whether the "80% rule" is being met, though these days we mention that less than we used to.

I would say in this case this being totally an *option*, it can stay safely in contrib. Core provides a satisfactory alternative to HTML5 Details and anyone who wants otherwise can get this in contrib.

**Comment by indigoxela** [comment]
@jlfranklin OK, you did a release of the contrib module. But it's currently not working properly. Can you please revoke that release until we're able to fix it? You made me maintainer - without asking and although it should be clear that I'd prefer to have it in core.

**Comment by jlfranklin** [comment]
> You made me maintainer.

Excluding you as one of the maintainers would come off as trying to snatch the work from you, which is absolutely not the case.  As requested, I've removed you from the maintainer list.

**Comment by indigoxela** [comment]
> Excluding you as one of the maintainers would come off as trying to snatch the work from you

The code is supposed to be GPLv2, so I see no problem if you "snatch" it. Supposing I'd volunteer to be a maintainer of a module I don't want at all is pretty weird, though. That's why I objected.

I still encourage people to test my PR, BTW. I didn't give up yet, I'm still convinced that the details element belongs to core form API.

Finding consensus can be hard, but a single person should not be able to completely block something several other people want. It'll just take longer to find the proper solution. That's not uncommon in the Open Source world.

**Comment by laryn** [comment]
In general I'm in agreement with the "modernize core as much as possible OOTB" line of thinking and including basic options for widely supported and standardized form elements seems in line with that, even if we can't replace all the core usages of collapsible fieldsets currently. 

So: I support adding the details element to the form API.

**Comment by klonos** [comment]
> So: I support adding the details element to the form API.

Me too 👍

**Comment by klonos** [comment]
That's right @laryn 👍🏼 ...we should add it now, and use it in newly-introduce things in core + allow contrib to use it. later, in Backdrop 2.0, we can consider replacing fieldsets with details elements throughout core, and deprecate fieldsets.

**Comment by indigoxela** [comment]
> in Backdrop 2.0, we can consider replacing fieldsets with details elements throughout core, and deprecate fieldsets.

More precisely, we can replace *collapsible* fieldsets with details, at least, that's how I understand it. The "normal" fieldset is OK and might still be valid for just grouping elements. But, agreed, replacing existing js-animated fieldsets is a 2.x task - for compatibility reasons.

**Comment by klonos** [comment]
> More precisely...

Precisely! 😅

**Comment by klonos** [comment]
Based on the good results that I had when I used the code from this PR for the PoC PR in #5090 (it was RTBC'ed pending the addition of the IE polyfill), I'm marking this as WFM. I'll have a look at the code, and hope that it's not too late for 1.20 🤞🏼

**Comment by indigoxela** [comment]
@klonos I had a hard time reverting the temporary changes, resolving conflicts and rebasing after all that time. The PR now should only change what it's supposed to change, and ship with everything it's supposed to ship with...

Can you verify? Might be a bit tricky.

This means, there's nothing to "test" here anymore, because nothing visible actually changes.

**Comment by jenlampton** [comment]
This PR looks really good! 

I found a few bits of CSS that I think we can remove, all targeting `details.filter-wrapper`. I believe the only place this appears in core is on the text-format selection for filtered text: "Formatting Options". Since we aren't using a details element there anymore (we're still using fieldsets) I don't think we need the special CSS to ensure those look good using a details element.

**Comment by docwilmot** [comment]
Also found a small nitpick. Otherwise the code is fine. Returning to needs work though,

**Comment by laryn** [comment]
I tried to make those minor CSS fixes. One more review would be good.

**Comment by laryn** [comment]
Thanks to @indigoxela for persisting on this one! Thanks to @jenlampton, @docwilmot, @klonos for reviewing and testing! And thanks to all for the spirited discussion that led to this scaled down version of the original PR, which sets us on a good path for the future.

Merged into 1.x for 1.20. Consider https://github.com/backdrop/backdrop-issues/issues/5090 unblocked! 

Re-opening, however, for a change record and documentation on the Form API page.

**Comment by indigoxela** [comment]
Thank you all for this last-minute-relay-race. :wink:  Great teamwork!

> ...documentation on the Form API page

@bugfolder might want to take over the API doc part.

**Comment by bugfolder** [comment]
Yes. Presumably docs should be updated after 1.20 is released.

**Comment by jenlampton** [comment]
@bugfolder change records can be published as soon as the feature gets in. There's a "changed in version" section where we add the version number so it's not necessary to wait for the release.  Here's the Drupal 8 change record I was going to copy/modify for Backdrop: https://www.drupal.org/node/1852020

**Comment by bugfolder** [comment]
I ws thinking @indigoxela's comment referred to updating this page:

https://docs.backdropcms.org/form_api

(Which needs a lot more work in bringing other bits up to par, including a bunch of outstanding PRs.)

I'm tied up for another several days at any rate, so I'll leave the change record to you.

**Comment by jenlampton** [comment]
> I ws thinking @indigoxela's comment referred to updating this page: docs.backdropcms.org/form_api

AH yes. We need that too :)

**Comment by jenlampton** [comment]
Change record: https://docs.backdropcms.org/change-records/a-new-form-api-details-element-has-been-added
